be/src/exec/pipeline/pipeline_driver_dispatcher.cpp

```diff
@@ -29,6 +29,16 @@ void GlobalDriverDispatcher::change_num_threads(int32_t num_threads) {
     }
 }
 
+void GlobalDriverDispatcher::finalize_driver(DriverPtr& driver, RuntimeState* runtime_state, DriverState state) {
+    DCHECK(driver);
+    driver->finalize(runtime_state, state);
+    if (driver->query_ctx()->is_finished()) {
+        auto query_id = driver->query_ctx()->query_id();
+        driver.reset();
+        QueryContextManager::instance()->remove(query_id);
```

Comment:
is all FragmentContexts destructed before this? I don't seem to see it.

Reply:
that assertion "driver->query_ctx()->is_finished()" is true indicates that all the FragmentContexts belonging to QueryContext has finished, since each FragmentContexts count down the atomic variable QueryContext::_num_active_fragments when it finishes, finally the QueryContext::_num_active_fragments reaches zero, at this moment, QueryContext::is_finished() return true.